DigitalLearningSolutions.Web/Helpers/RegistrationEmailValidator.cs

```diff
@@ -0,0 +1,99 @@
+            PersonalInformationViewModel model,
+            ModelStateDictionary modelState,
+            IUserService userService
+        )
+        {
+            ValidateEmailAddresses(false, model, modelState, userService);
+        }
+
+        public static void ValidateEmailAddressesForAdminRegistration(
+            PersonalInformationViewModel model,
+            ModelStateDictionary modelState,
+            IUserService userService,
+            ICentresService centresService
+        )
+        {
+            ValidateEmailAddresses(true, model, modelState, userService, centresService);
+        }
+
+        private static void ValidateEmailAddresses(
+            bool isRegisterAdminJourney,
+            PersonalInformationViewModel model,
+            ModelStateDictionary modelState,
+            IUserService userService,
+            ICentresService? centresService = null
+        )
+        {
+            var primaryEmailIsValidAndNotNull =
+                !modelState.HasError(nameof(PersonalInformationViewModel.PrimaryEmail)) &&
+                model.PrimaryEmail != null;
+            var secondaryEmailIsValidAndNotNull =
+                !modelState.HasError(nameof(PersonalInformationViewModel.SecondaryEmail)) &&
+                model.SecondaryEmail != null;
+
+            var primaryEmailMatchesCentre = centresService?.DoesEmailMatchCentre(
+                model.PrimaryEmail!,
+                model.Centre!.Value
+            );
+            var secondaryEmailMatchesCentre = centresService?.DoesEmailMatchCentre(
+                model.SecondaryEmail!,
+                model.Centre!.Value
+            );
+
+            if (primaryEmailIsValidAndNotNull)
+            {
+                if (userService.EmailIsInUse(model.PrimaryEmail))
+                {
+                    modelState.AddModelError(
+                        nameof(PersonalInformationViewModel.PrimaryEmail),
+                        DuplicateEmailErrorMessage
+                    );
+                }
+                else if (isRegisterAdminJourney && model.SecondaryEmail == null &&
+                         primaryEmailMatchesCentre! == false)
+                {
+                    modelState.AddModelError(
+                        nameof(PersonalInformationViewModel.PrimaryEmail),
+                        WrongEmailForCentreErrorMessage
+                    );
+                }
+            }
+
+            if (secondaryEmailIsValidAndNotNull)
+            {
+                if (userService.EmailIsInUse(model.SecondaryEmail))
+                {
+                    modelState.AddModelError(
+                        nameof(PersonalInformationViewModel.SecondaryEmail),
+                        DuplicateEmailErrorMessage
+                    );
+                }
+                else if (isRegisterAdminJourney && secondaryEmailMatchesCentre! == false &&
+                         primaryEmailMatchesCentre! == false)

```

Comment:
I don't think we need the `!` operator on `primaryEmailMatchesCentre`, nor on the `secondaryEmailMatchesCentre` in these comparisons in this method.

Reply:
removed!